### ISSUE TITLE
Inaccessible NextCloud instance due to missing login form after upgrade to NC17

### ISSUE THREAD

**Issue by didierm** [nextcloud/server#18101]
After upgrading one of our instances (labelled "Sandbox") to v17.0.1, the login form ceases to appear, rendering our instance completely inaccessible.

To illustrate the issue, two screenshots (including the respective browser logs) are included : one from the inaccessible "Sandbox" instance, and - for comparative purposes - one from a working v16.0.4 instance (labelled "Private").

* Tested with different browsers on several platforms.
* We do not currently use SAML.

### Steps to reproduce
1. Upgrade v16.0.4 -> v17.0.1
2. Observe loss of login form
3. Weep

### Expected behaviour
Login form present

### Actual behaviour
Login form absent

### Server configuration

**Operating system**: CentOS Linux release 7.7.1908 (Core)

**Web server:** Apache 2.4.6

**Database:** PostgreSQL 9.2.24

**PHP version:** PHP 7.2.25

**Nextcloud version:** version: 17.0.1.1

**Updated from an older Nextcloud/ownCloud or fresh install:** upgrade from 16.0.4.1

**Where did you install Nextcloud from:**

**Signing status:**
<details>
<summary>Signing status</summary>

```
integrity OK
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
Enabled:
  - accessibility: 1.3.0
  - activity: 2.10.1
  - admin_audit: 1.7.0
  - bruteforcesettings: 1.4.0
  - calendar: 1.7.1
  - cloud_federation_api: 1.0.0
  - comments: 1.7.0
  - contacts: 3.1.6
  - dav: 1.13.0
  - federatedfilesharing: 1.7.0
  - federation: 1.7.0
  - files: 1.12.0
  - files_antivirus: 2.2.0
  - files_downloadactivity: 1.6.0
  - files_external: 1.8.0
  - files_pdfviewer: 1.6.0
  - files_rightclick: 0.15.1
  - files_sharing: 1.9.0
  - files_trackdownloads: 1.6.0
  - files_videoplayer: 1.6.0
  - firstrunwizard: 2.6.0
  - gallery: 18.4.0
  - impersonate: 1.4.0
  - logreader: 2.2.0
  - lookup_server_connector: 1.5.0
  - nextcloud_announcements: 1.6.0
  - notifications: 2.5.0
  - oauth2: 1.5.0
  - password_policy: 1.7.0
  - polls: 0.10.4
  - privacy: 1.1.0
  - provisioning_api: 1.7.0
  - recommendations: 0.5.0
  - richdocuments: 3.4.4
  - serverinfo: 1.7.0
  - sharebymail: 1.7.0
  - sharepoint: 1.5.0
  - survey_client: 1.5.0
  - systemtags: 1.7.0
  - tasks: 0.11.3
  - text: 1.1.1
  - twofactor_backupcodes: 1.6.0
  - updatenotification: 1.7.0
  - user_ldap: 1.7.0
  - viewer: 1.2.0
  - workflowengine: 1.7.0
Disabled:
  - circles
  - encryption
  - files_trashbin
  - files_versions
  - jsloader
  - ownpad
  - ransomware_detection
  - support
  - theming
```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "cloud2.irc.ugent.be",
            "cp0582.irc.ugent.be",
            "cloud.irc.ugent.be",
            "cp0488.irc.ugent.be"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "version": "17.0.1.1",
        "dbtype": "pgsql",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "dbtableprefix": "nc_",
        "installed": true,
        "skeletondirectory": "",
        "lost_password_link": "disabled",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpmode": "smtp",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_template_class": "IRC\\IRCTheme",
        "overwritehost": "cloud.irc.ugent.be",
        "overwriteprotocol": "https",
        "overwritewebroot": "\/sandbox",
        "overwrite.cli.url": "https:\/\/cloud.irc.ugent.be\/sandbox",
        "trashbin_retention_obligation": "auto",
        "connectivity_check_domains": [],
        "updatechecker": false,
        "updater.release.channel": "stable",
        "log_type": "file",
        "logfile": "\/srv\/nextcloud\/cloud00\/cloud00.log",
        "loglevel": 0,
        "logtimezone": "Europe\/Brussels",
        "log_query": false,
        "log_rotate_size": 33554432,
        "apps_paths": [
            {
                "path": "\/opt\/nextcloud.staging\/apps",
                "url": "\/apps",
                "writable": false
            },
            {
                "path": "\/srv\/nextcloud\/cloud00\/apps",
                "url": "\/apps-appstore",
                "writable": true
            }
        ],
        "enable_previews": false,
        "preview_libreoffice_path": "\/usr\/bin\/libreoffice",
        "sort_groups_by_name": true,
        "maintenance": false,
        "memcache.local": "\\OC\\Memcache\\APCu",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 0
        },
        "tempdirectory": "\/srv\/nextcloud\/tempdirectory\/cloud00\/temp",
        "cache_path": "\/srv\/nextcloud\/tempdirectory\/cloud00\/cache",
        "theme": "cloud00",
        "filesystem_check_changes": 1,
        "secret": "***REMOVED SENSITIVE VALUE***",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "upgrade.disable-web": true,
        "debug": true,
        "userdir_readonly": true,
        "ldapIgnoreNamingRules": false,
        "ldapProviderFactory": "OCA\\User_LDAP\\LDAPProviderFactory"
    }
}
```
</details>

**Are you using external storage, if yes which one:** local/smb/nfs

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** LDAP

#### LDAP configuration (delete this part if not used)
<details>
<summary>LDAP config</summary>

```
disabling LDAP app "user_ldap" yields no improvement
```
</details>

### Client configuration
**Browser:** all

**Operating system:** all

#### Web server error log
<details>
<summary>Web server error log</summary>

```
{ip_removed} - - [25/Nov/2019:11:24:47 +0100] "GET /sandbox/index.php/login HTTP/1.1" 200 13118
{ip_removed} - - [25/Nov/2019:11:24:47 +0100] "GET /sandbox/apps/files_pdfviewer/css/style.css HTTP/1.1" 200 346
{ip_removed} - - [25/Nov/2019:11:24:47 +0100] "GET /sandbox/core/css/guest.css HTTP/1.1" 200 19902
{ip_removed} - - [25/Nov/2019:11:24:47 +0100] "GET /sandbox/themes/cloud00/core/css/server.css HTTP/1.1" 200 3635
{ip_removed} - - [25/Nov/2019:11:24:47 +0100] "GET /sandbox/core/js/dist/main.js HTTP/1.1" 200 1300833
{ip_removed} - - [25/Nov/2019:11:24:47 +0100] "GET /sandbox/core/js/mimetype.js HTTP/1.1" 200 3654
{ip_removed} - - [25/Nov/2019:11:24:47 +0100] "GET /sandbox/core/js/dist/main.js.map HTTP/1.1" 200 4529843
{ip_removed} - - [25/Nov/2019:11:24:47 +0100] "GET /sandbox/core/js/mimetypelist.js HTTP/1.1" 200 6119
{ip_removed} - - [25/Nov/2019:11:24:47 +0100] "GET /sandbox/core/js/select2-toggleselect.js HTTP/1.1" 200 1479
{ip_removed} - - [25/Nov/2019:11:24:47 +0100] "GET /sandbox/core/search/js/search.js HTTP/1.1" 200 5560
{ip_removed} - - [25/Nov/2019:11:24:47 +0100] "GET /sandbox/core/js/dist/share_backend.js HTTP/1.1" 200 117299
{ip_removed} - - [25/Nov/2019:11:24:47 +0100] "GET /sandbox/apps/files_pdfviewer/js/previewplugin.js HTTP/1.1" 200 6000
{ip_removed} - - [25/Nov/2019:11:24:47 +0100] "GET /sandbox/apps/files_videoplayer/js/main.js HTTP/1.1" 200 6344
{ip_removed} - - [25/Nov/2019:11:24:48 +0100] "GET /sandbox/core/search/js/searchprovider.js HTTP/1.1" 200 12444
{ip_removed} - - [25/Nov/2019:11:24:48 +0100] "GET /sandbox/core/js/files/fileinfo.js HTTP/1.1" 200 2142
{ip_removed} - - [25/Nov/2019:11:24:48 +0100] "GET /sandbox/core/js/files/client.js HTTP/1.1" 200 23663
{ip_removed} - - [25/Nov/2019:11:24:48 +0100] "GET /sandbox/core/js/dist/login.js HTTP/1.1" 200 648867
{ip_removed} - - [25/Nov/2019:11:24:48 +0100] "GET /sandbox/core/img/loading-dark.gif HTTP/1.1" 200 4683
{ip_removed} - - [25/Nov/2019:11:24:48 +0100] "GET /sandbox/core/img/actions/toggle.svg HTTP/1.1" 200 308
{ip_removed} - - [25/Nov/2019:11:24:48 +0100] "GET /sandbox/themes/cloud00/core/img/logo-cloud00.png HTTP/1.1" 200 36801
{ip_removed} - - [25/Nov/2019:11:24:48 +0100] "GET /sandbox/index.php/apps/accessibility/js/accessibility?v=0 HTTP/1.1" 200 65
{ip_removed} - - [25/Nov/2019:11:24:48 +0100] "GET /sandbox/core/img/favicon-touch.png HTTP/1.1" 200 2918
{ip_removed} - - [25/Nov/2019:11:24:48 +0100] "GET /sandbox/themes/cloud00/core/img/favicon-cloud00.ico HTTP/1.1" 200 3262
{ip_removed} - - [25/Nov/2019:11:24:48 +0100] "GET /sandbox/index.php/csrftoken HTTP/1.1" 200 101

```
</details>

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>
'debug' => true,
'loglevel' => 0,
'log_query' => true,

...
Debug    core               DB QueryBuilder: 'SELECT `id`, `uid`, `login_name`, `password`, `name`, `token`, `type`, `remember`, `last_activity`,    2019-11-25T08:58:02+01:00 
						  `last_check`, `scope`, `expires`, `version` FROM `*PREFIX*authtoken` WHERE (`token` = :dcValue1) AND (`version` =                                  
						  :dcValue2)' with parameters: dcValue1 =>                                                                                                           
						  'c356{removed}a832',                           
						  dcValue2 => '1'                                                                                                                                    

Debug    core               DB QueryBuilder: 'SELECT * FROM `*PREFIX*authtoken` WHERE (`token` = :dcValue1) AND (`version` = :dcValue2)' with        2019-11-25T08:58:02+01:00 
						  parameters: dcValue1 =>                                                                                                                            
						  'c356{removed}a832',                           
						  dcValue2 => '2'    
...

(note : query logs taken at an earlier attempt, hence timestamps are not matching with webserver logs)
</details>

#### Browser log
<details>
<summary>Browser log</summary>
```
GET https://cloud.irc.ugent.be/sandbox/core/js/dist/login.js [HTTP/1.1 200 OK 7ms]

Could not find initial state loginErrors of core initialstate.js:31:10
Could not find initial state loginMessages of core initialstate.js:31:10
Could not find initial state loginRedirectUrl of core initialstate.js:31:10
Could not find initial state loginUsername of core initialstate.js:31:10
Could not find initial state loginThrottleDelay of core initialstate.js:31:10
Could not find initial state loginCanResetPassword of core initialstate.js:31:10
Could not find initial state loginResetPasswordLink of core initialstate.js:31:10
Could not find initial state loginAutocomplete of core initialstate.js:31:10
Could not find initial state resetPasswordTarget of core initialstate.js:31:10
Could not find initial state resetPasswordUser of core initialstate.js:31:10
```
For full, comparative browser logs, see screenshots.

</details>

**Comment by didierm** [comment]
Version 16.0.4 with login form : 

![image](https://user-images.githubusercontent.com/989004/69536149-bfb72180-0f7d-11ea-884b-5c049512e078.png)

**Comment by didierm** [comment]
Version 17.0.1 without login form :

![image](https://user-images.githubusercontent.com/989004/69536200-d8bfd280-0f7d-11ea-9a44-d0fefc2c7be6.png)

**Comment by didierm** [comment]
I looked at #17487 , as that issue also reports 

`Could not find initial state loginErrors of core initialstate.js` , etc.

but we are not currently using SSO/SAML , and the fix https://github.com/nextcloud/server/pull/17548 should already be in 17.0.1.

**Comment by wdfee** [comment]
Did you upgrade via occ command or via web updater? In my case I used the web updater and had a similar screen, I had to continue the upgrade process via command line, this worked for me.

**Comment by didierm** [comment]
@wdfee Thank you for your feedback.

We always upgrade by means of the `occ upgrade -v` command, as this provides the most valuable feedback. The upgrade process finsihed without errors.

**Comment by kesselb** [comment]
> "theme": "cloud00",

I guess you have to update the theme. Login uses Vue.js with Nextcloud 17.

**Comment by didierm** [comment]
@kesselb Thanks ! (we were already looking in the direction of Vue.js, with no immediate results)

As we indeed use a custom theme, part of our upgrade protocol compares the standard themes of the running version (in this case : v16.0.4.1) with the new version (v17.0.1.1).
When upgrading, we found no substantial differences between NC16 and NC17 (apart from a small version number update).

As Google-Fu doesn't reveal anything in particular, do you care to provide an additional pointer concerning Vue.js and NC17 theming ?

**Comment by kesselb** [comment]
https://github.com/nextcloud/server/blob/stable17/core/templates/login.php
https://github.com/nextcloud/server/commit/64c4bb5bce52599d9df4007ba9d622840fee1213#diff-c757840e10799bae317eae4ed1701009

**Comment by didierm** [comment]
@kesselb Thanks.

Not sure how to continue from here ; the latest documented version w.r.t. theming appears to be NC15 : https://docs.nextcloud.com/server/15/developer_manual/core/theming.html .

**Comment by kesselb** [comment]
> the latest documented version w.r.t. theming appears to be NC15 

https://github.com/nextcloud/documentation/pull/946

> Not sure how to continue from here

What do you mean? If you overwrite the login page with your theme update the markup ;) I guess Vue.js is looking for `<div id="login"></div>` to render the login form. Don't forget to include dist/login.

**Comment by didierm** [comment]
@kesselb Thank you for your assistance.

This issue was caused by a local (NC12 era) `{custom_theme}/core/templates/layout.guest.php` file, which was compatible up to NC16, but proved incompatible with NC17.

We use the `layout.guest.php` template overriding the standard `core/templates/layout.guest.php`, in order to be able to provide a custom (themed) favicon on the NC login page instead of the default one, which is hard-coded : 
https://github.com/nextcloud/server/blob/3eb3c3f884e1791059e71f5106880ea507f5c12f/core/templates/layout.guest.php#L18

The hardcoded path is probably the cause of https://github.com/nextcloud/server/issues/16504 (closed by author, but we confirm/observe this in NC17).


We have found 2 alternatives : 

1. Do not hardcode the favicon path, but respect/allow for theming, e.g. : 
```
<link rel="icon" href="<?php p($theme->getIcon()); /* IE11+ supports png */ ?>">
```
This requires additional changes in `lib/public/Defaults.php` and `lib/private/legacy/defaults.php`

2. 
Use the theming app (which has a favicon setting).
Unfortunately, there are other issues with the themin app (i.e. deconstructing the themed footer with copyright and entity, etc.).
